### ISSUE TITLE
Luck based effects on card dealing.

### ISSUE THREAD

**Issue by TheWhiteCollarPlayers** [Scrivener07/FO4_Gambling#3]
I have been discussing with Shadow from the Cascadia team about general ways to implement a Luck based effect on the "randomness" of dealing cards to the player. I.E. at higher luck levels you will get dealt better cards. I will copy over his suggestions, I am thinking that certain degrees of combinations of his suggestions might work best in order to achieve a more RPG feel to gambling.

**Comment by TheWhiteCollarPlayers** [comment]
From shadow:
1-2 is considered "low int" in NV with the stupid dialogue options, so 1-2 is bad luck, as an assumption
so make their first card random
but then figure out how far they are from busting/blackjack
and deal their second card in a set of ranges based on luck
higher luck gets lower cards/higher chance of hitting 21
low/bad luck gets higher cards/higher chance of busting.

You may also not want to do that on every turn
So some hands for the player are truly random(edited)
but you could work in that system that works off of luck instead too, and throw that one to the player
If Player.GetValue(Luck) <= Utility.RandomInt(1, 50)
    ; luck based hand
    
Else
    ; random hand
    
EndIf

1 in 5 chance with maxed luck, 1 in 50 chance with min luck
Of couse RandomInt(1,50) could be tweaked too
range could be 1-100 for 10% chance with 10 luck
or lower for higher chances
You could even calculate the max range for randomint off of luck

**Comment by Scrivener07** [comment]
I agree we should factor in the players luck attribute to the odds of winning a game. The technical details are a little off base though. The card that gets drawn from a deck is much more than simply a random number or an index in a list so just returning a random number wont do the trick.

Lets recap the goal here. We want to increase the players chance of winning a game when their luck is high and decrease the chance of winning when their luck is low. To achieve this we should do it exactly the same way it would be done in real life. In real life you would decrease the odds of winning in favor of the house by stacking the deck with multiple decks of cards. Since the `Deck` class models real life as closely as I could we can add more card entries to the decks internal card array which will decrease the chances of the player winning a game.

Here is a chart of the odds based on the amount of decks used in a game of black jack.

| Number of Decks | House Advantage |
| ------------------- |:--------------------:|
| Single Deck          | 0.17% |
| Double Deck        | 0.46% |
| Four Decks           | 0.60% |
| Six Decks              | 0.64% |
| Eight Decks          | 0.65% |

With that, we increase the amount of decks used the lower a players luck attribute is. When a game uses a single deck of 52 cards the player should have a luck of 10.

So I propose a formula like this.
`(10 - LUCK) + 1 = DecksUsed`

| Luck | Decks |
| ----- |:-------:|
|  1      | 10 |
|  2      |  9  |
|  3      |  8  |
|  4      |  7  |
|  5      |  6  |
|  6      |  5  |
|  7      |  4  |
|  8      |  3  |
|  9      |  2  |
|  10    |  1  |

**Comment by TheWhiteCollarPlayers** [comment]
That's genius! And very realistic as well, I love it! But...how smoothly can you implement that? Of course, a simple way would be to run a set of comparisons against the player's Luck stat even before the cards are "shuffled" then add decks based on the stat.

**Comment by Scrivener07** [comment]
Very smoothly if I may toot my own horn. Off the top of my head I can expose like a `int Depth` property on the `Deck` class. Then when `NewCards` is called it can add 52 card deck x times. That could potentially be up to 520 entries in the internal array so I will need to find a workaround for the 128 element array limit. Since this is a predictable maximum using multiple internal arrays may be a good option.

[Deck Class](https://github.com/Scrivener07/FO4_Gambling/blob/a52625a96fe3ff85416570359b4f787d7e335b2e/Data/Scripts/Source/FO4_Gambling/Gambling/Deck.psc)

**Comment by TheWhiteCollarPlayers** [comment]
Wasn't aware of that limit, ugh if this were C I'd say use a pointer in the last element of the array to point to the next array...but alas. In that case, multiple internal arrays would work (pretty sure I know what means).

**Comment by TheWhiteCollarPlayers** [comment]
I do intend to do this, however this might be best saved for when we want to revisit the blackjack code for optimization and improvements. The source code for other games will incorporate the luck stat in some way

**Comment by Scrivener07** [comment]
During our private discussion we talked about making this simple. For each point of luck, grant a bonus to the amount of caps a game will payout.

**Comment by TheWhiteCollarPlayers** [comment]
Leaving this open, one of those things we can do quickly, a simple check for the multiplier on the payout should suffice for blackjack

**Comment by Scrivener07** [comment]
The user **[ForsakenShell](https://github.com/ForsakenShell)** had some thoughts to add to the conversation that I found interesting.

> I was suggest using a logarithmic scale so as your luck gets higher it effects game outcomes less and less.
> 
> **Example:** `Log( 10 ) ^Luck`
> When it hits 7-8 luck it starts really slowing down, after 10 it crawls.
> 

> Produces a range of 0-10 for luck 1-10 where `1 = 0`, `5 = 0.7`; `7 = 0.85`, `10 = 1`.
> Going beyond 10 to 20 only yields `1.3`.
> I didn't hit any snags with the mod yet though but I didn't get to test it for long either.

_Additional Testing_
> **Example:**  `5 * Log( 10 ) ^ Luck`
> Think of the last provided equation as "% chance the player will get a favourable card"
> At 10 luck it yields `5%`. 
> At 20 luck is only ~`6.5%`. 
> At 1 luck it is `0%`.


Thanks for adding to the conversation **E**.

**Comment by Scrivener07** [comment]
**User Feedback**
![capture1](https://user-images.githubusercontent.com/1912967/37377644-81328e56-2701-11e8-804d-4f541d98f27f.PNG)

**User Feedback**
![capture](https://user-images.githubusercontent.com/1912967/37377645-839fb9ca-2701-11e8-86c7-c71a72769647.PNG)
